File: tests/ps2.js
var iter = 0;
var id = 0;

// NEED TO SET THIS UP DYNAMICALLY
var numWorkers = 3;
var workersReady = [1, 1, 1];

jasync function printamsg(x) {
    console.log("Got a message from the worker ", x);
}

jsync function getID() {
    // if(id == 0){
    //     id++;
    //     return 0;
    // }
    
    // id++;
    return id++;
}

// function to initialize stuff at the beginning
// setTimeout(function(){
//     // workersReady = new Array(numWorkers); // This seems disfunctional
//     workersReady = [0, 0, 0];
//     for(var i = 0; i < numWorkers; i++) {
//         workersReady[i] = 1;
//     }
//     console.log("Initialized the workers' array.");
// }, 0)

var rv;

setInterval(function() {
    
    if(!isNextWave()) {
        return;
    }
    //workersReady = [0, 0, 0];

    console.log("Calling a synchronous function... in C side");
    rv = runatsametime(iter);
    iter++;
    console.log("Returns ", rv);
    //console.log("Worker's Readiness : ", workersReady[0], workersReady[1], workersReady[2]);
}, 5000);

// ------ sync functions ------ //

jsync function wReady(x) {
//    workersReady[x] = 1;
    //    return workersReady[x];
    return 1;
}

function isNextWave() {
    var isReady = true;
    for(var i = 0; i < numWorkers; i++) {
        if(workersReady[i] == 0) {
            isReady = false;
        }
    }
    return isReady;
}

File: tests/ps2.c
#include <stdio.h>

int getID();
int wReady(int x);

int cnt = 0;
int id = -1;

int xxxx() {
    return 1;
}


jsync int runatsametime(int x) {

    printf("Run the barrier function at step %d from worker #%d\n", x, id);
    //localworker(x);
    int temp = 0;
    while(!temp) {
//        temp = wReady(id);
	temp = wReady(1);
	
    } 
    cnt++;
    return cnt;
}


// jasync localworker(int x) {
//     //cnt++;
//     printf("Prints from the local worker from the %d-th barrier.\n", x);
//     jsleep(100);
// }

jasync get_myid() {
  while(1) {
    id = getID();
    if (id >= 0) {
        printf("Got my id --> %d", id);
        return;
    }
    jsleep(300);
  }
}

jasync stupidtask() {
    while(1) {
         printf("Doing pointless stuff...\n");
         jsleep(1000);
    }
}

int main(int argc, char *argv[]) {
    get_myid();
    stupidtask();
    return 0;
}
